File: app/models/vaccination.py
from sqlalchemy import Column, Integer, String, ForeignKey, Date, Boolean, Enum as SQLEnum, Index
from .base import Base
import enum

class VaccinationStatus(str, enum.Enum):
    SCHEDULED = "scheduled"
    COMPLETED = "completed"
    POSTPONED = "postponed"

class Vaccination(Base):
    __tablename__ = "vaccinations"

    id = Column(Integer, primary_key=True, autoincrement=True, index=True)
    baby_id = Column(Integer, ForeignKey("babies.id", ondelete="CASCADE"), nullable=False)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    
    vaccine_name = Column(String, nullable=False)
    dose_number = Column(Integer, nullable=False)
    status = Column(SQLEnum(VaccinationStatus), nullable=False, default=VaccinationStatus.SCHEDULED)
    
    scheduled_date = Column(Date, nullable=False, index=True)
    completed_date = Column(Date, nullable=True)
    
    lot_number = Column(String, nullable=True)
    hospital_name = Column(String, nullable=True)
    has_side_effect = Column(Boolean, nullable=False, default=False)
    notes = Column(String, nullable=True)

    __table_args__ = (
        Index("idx_vaccination_baby_status", "baby_id", "status"),
    )

File: frontend/schemas/vaccination.ts
import { z } from "zod"

export const vaccinationSchema = z.object({
  vaccine_name: z.string().min(1, "ワクチン名を入力してください"),
  dose_number: z.coerce.number().min(1, "回数は1以上を指定してください"),
  status: z.enum(["scheduled", "completed", "postponed"]),
  scheduled_date: z.string().min(1, "予定日を選択してください"),
  completed_date: z.string().optional().nullable(),
  lot_number: z.string().optional().nullable(),
  hospital_name: z.string().optional().nullable(),
  has_side_effect: z.boolean().default(false),
  notes: z.string().optional().nullable(),
})

export type VaccinationFormValues = z.infer<typeof vaccinationSchema>
